### PR TITLE
:sparkles: Markers can now indicate their priority when applying

### PR DESCRIPTION
**Pull request by dprotaso**
There's a need for more control over when SchemaMarkers are applied to a Schema. Right now there are two levels - 'First' and everything else. The iteration over the markers within this level is non-deterministic because we're iterating over a golang map type.

As an alternative we now collect all the markers and sort them based on a new priority. Markers can signal their priority by implementing the optional interface

```go

type ApplyPriority int64

type ApplyPriorityMarker interface {
 	ApplyPriority() ApplyPriority
 }
```

This change is backwards compatible - downstream markers that implement `ApplyFirst()` will behave the same as before.

### CODE REVIEW

**Comment by dprotaso** [comment]
Hey folks, just following up - is there anything else needed for this PR?

File: pkg/crd/markers/doc.go [hunk 1]
```diff
@@ -25,8 +25,8 @@ limitations under the License.
 // (crd.SchemaMarker).  Any marker implementing this will automatically
 // be run after the rest of a given schema node has been generated.
 // Markers that need to be run before any other markers can also
-// implement ApplyFirst, but this is discouraged and may change
-// in the future.
+// implement ApplyFirst or markers.ApplyPriority, but this is discouraged
+// and may change in the future.
```
Comment (camilamacedo86) [inline pkg/crd/markers/doc.go:29]: Could we add here an example over how to use both?
Beyond that all fine 🥇 
Great work

**Comment by dprotaso** [comment]
@camilamacedo86 done - take a look

File: pkg/crd/markers/doc.go [hunk 2]
```diff
@@ -26,20 +26,32 @@ limitations under the License.
...
 //
+//  type MyCustomMarker string
+//
+//  func (m MyCustomMarker) ApplyPriority() ApplyPriority {
+//  	return ApplyPriorityFirst
+//  }
+//
+//  func (m MyCustomMarker) ApplyToSchema(schema *apiext.JSONSchemaProps) error {
+//    ...
+//  }
```
Comment (camilamacedo86) [inline pkg/crd/markers/doc.go:41]: Could we change it for. something like

interface in combination with ApplyPriorityDefault and
// ApplyPriorityFirst constants. Following an example over how to use them:

**Review by camilamacedo86 — COMMENTED** [review]
I am ok with but would be nice get thoughts from others like @alvaroaleman @varshaprasad96 

Otherwise, lgtm

File: pkg/crd/markers/doc.go (hunk 2 above)
Comment (dprotaso) [reply pkg/crd/markers/doc.go]: Done

**Comment by dprotaso** [comment]
Hi - folks just following up

**Review by camilamacedo86 — COMMENTED** [review]
It seems good or me. 🥇 
Let's see what others think about

/lgtm

**Comment by dprotaso** [comment]
bump @vincepri @alvaroaleman

File: pkg/crd/markers/priority.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package markers
+
+import "math"
+
+type ApplyPriority int64
```
Comment (vincepri) [inline pkg/crd/markers/priority.go:21]: Add a comment on what this type does?

File: pkg/crd/markers/priority.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+const (
+	// ApplyPriorityDefault is the default priority for markers
+	// that don't implement ApplyPriorityMarker
+	ApplyPriorityDefault ApplyPriority = math.MaxInt64 >> 2
+
+	// ApplyPriorityFirst is the priority value assigned to markers
+	// that implement the ApplyFirst() method
+	ApplyPriorityFirst ApplyPriority = ApplyPriorityDefault - 10
```
Comment (vincepri) [inline pkg/crd/markers/priority.go:30]: This is a bit hard to understand imo, could we consider an alternative where we define constants that are clear on where they're used.

For example:
- `ApplyPriorityFirst` = 1
- `ApplyPriorityTypes` = 5
-  `ApplyPriorityDefault` = 10

And so on?

**Comment by dprotaso** [comment]
Seems like prow is broken?

/retest

**Comment by dprotaso** [comment]
Prow started working again

@vincepri and/or @alvaroaleman PTAL

**Comment by dprotaso** [comment]
@vincepri and/or @alvaroaleman PTAL

**Comment by dprotaso** [comment]
@vincepri @alvaroaleman can you take a look please?

File: pkg/crd/schema.go [hunk 5]
```diff
@@ -126,39 +127,44 @@ func infoToSchema(ctx *schemaContext) *apiext.JSONSchemaProps {
 
 // applyMarkers applies schema markers to the given schema, respecting "apply first" markers.
 func applyMarkers(ctx *schemaContext, markerSet markers.MarkerValues, props *apiext.JSONSchemaProps, node ast.Node) {
+	markers := make([]SchemaMarker, 0, len(markerSet))
+
 	// apply "apply first" markers first...
```
Comment (sbueringer) [inline pkg/crd/schema.go:132]:
```suggestion
```

Seems oudated

File: pkg/crd/schema.go [hunk 6]
```diff
@@ -126,39 +127,44 @@ func infoToSchema(ctx *schemaContext) *apiext.JSONSchemaProps {
 
 // applyMarkers applies schema markers to the given schema, respecting "apply first" markers.
```
Comment (sbueringer) [inline pkg/crd/schema.go:128]: Let's please update this comment to match the new implementation

File: pkg/crd/schema_test.go [hunk 7]
```diff
@@ -110,3 +111,58 @@ func Test_Schema_MapOfStringToArrayOfFloat32(t *testing.T) {
...
+			&testapplyFirstMarker{callback: func() {
+				invocations = append(invocations, "applyFirst")
+			}},
+			&testPriorityMarker{priority: crdmarkers.ApplyPriorityDefault, callback: func() {
+				invocations = append(invocations, "default")
+
+			}},
+		}}, props, nil)
+
+	g.Expect(invocations).To(gomega.Equal([]string{"0", "1", "applyFirst", "default"}))
```
Comment (sbueringer) [inline pkg/crd/schema_test.go:143]: I think "1" and "applyFirst" are not guaranteed to be in that order. Both end up with priority 1 and right now we are only using sort.Slice.
```go
// The sort is not guaranteed to be stable: equal elements
// may be reversed from their original order.
// For a stable sort, use SliceStable.
```

I assume we don't want to guarantee an order for the same priority. So using sort.Slice should be fine. I just wonder if we should use distinct priorities for the test here

File: pkg/crd/schema_test.go [hunk 8]
```diff
@@ -110,3 +111,58 @@ func Test_Schema_MapOfStringToArrayOfFloat32(t *testing.T) {
...
+				},
+			},
+			&testPriorityMarker{priority: 1, callback: func() {
+				invocations = append(invocations, "1")
+
+			}},
+			&testapplyFirstMarker{callback: func() {
+				invocations = append(invocations, "applyFirst")
+			}},
+			&testPriorityMarker{priority: crdmarkers.ApplyPriorityDefault, callback: func() {
```
Comment (sbueringer) [inline pkg/crd/schema_test.go:137]: Would it make sense to use a different priority here and add an additional marker that fallsback to the default?

File: pkg/crd/markers/priority.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	// that don't implement ApplyPriorityMarker
+	ApplyPriorityDefault ApplyPriority = 10
+
+	// ApplyPriorityFirst is the priority value assigned to markers
+	// that implement the ApplyFirst() method
+	ApplyPriorityFirst ApplyPriority = 1
+)
+
+// ApplyPriorityMarker designates the order markers should be applied.
+// Lower priority indicates it should be applied first
```
Comment (sbueringer) [inline pkg/crd/markers/priority.go:34]: Should we mention that this only applies to Validation markers?

**Review by sbueringer — COMMENTED** [review]
Thx looks good. Just a few minor findings otherwise good to go from my side

File: pkg/crd/schema_test.go (hunk 7 above)
Comment (dprotaso) [reply pkg/crd/schema_test.go]: I added a new marker to test the default priority. Likewise I updated the test so all the priorities are unique.

File: pkg/crd/schema_test.go (hunk 8 above)
Comment (dprotaso) [reply pkg/crd/schema_test.go]: done

File: pkg/crd/markers/priority.go (hunk 9 above)
Comment (dprotaso) [reply pkg/crd/markers/priority.go:34]: ApplyPriority can apply to any marker - not just validation

Comment (sbueringer) [reply pkg/crd/markers/priority.go:34]: I was looking at the doc and there we distinguish between "Validation Markers" (implementing crd.SchemaMarker) and "CRD Markers" (implementing crd.SpecMarker). I only saw ApplyPriorityMarker being handled for SchemaMarker.

I might be missing something, but looks like it's not implemented for SpecMarker

File: pkg/crd/markers/doc.go [hunk 10]
```diff
@@ -19,27 +19,40 @@ limitations under the License.
...
+//	}
 //
 // All validation markers start with "+kubebuilder:validation", and
 // have the same name as their type name.
 //
-// CRD Markers
+// # CRD Markers
 //
 // Markers that modify anything in the CRD itself *except* for the schema
 // implement ApplyToCRD (crd.CRDMarker).  They are expected to detect whether
```
Comment (sbueringer) [inline pkg/crd/markers/doc.go:50]:
```suggestion
// implement ApplyToCRD (crd.SpecMarker).  They are expected to detect whether
```

Just saw when looking at this again that this comment is wrong. Would be nice if you can fix it :)

File: pkg/crd/markers/priority.go (hunk 9 above)
Comment (dprotaso) [reply pkg/crd/markers/priority.go:34]: Good catch - yeah it's not in the spec marker - I'll update the comment

**Comment by sbueringer** [comment]
Thank you!

/lgtm

/assign @alvaroaleman @vincepri

**Comment by sbueringer** [comment]
/approve

Thank you very much!

**Pull request merged**